GitHub Action Deploy TRE fails on first run at ACR login step

**Describe the bug**

GitHub Action Deploy Azure TRE fails on first run at the [2nd ACR Login step](https://github.com/microsoft/AzureTRE/blob/6d589c49d6d647633109c70ee4636415b1ac17a8/.github/workflows/deploy_tre_reusable.yml#L247) with error `$ec numeric argument required`.

![first-run](https://github.com/microsoft/AzureTRE/assets/7916532/4bdf3050-cfc9-4fd5-b530-1605a6f635a5)

On subsequent runs it passes, as the 2nd ACR Login step is not executed, due to the [1st ACR Login step](https://github.com/microsoft/AzureTRE/blob/6d589c49d6d647633109c70ee4636415b1ac17a8/.github/workflows/deploy_tre_reusable.yml#L201) passing:

![second-run](https://github.com/microsoft/AzureTRE/assets/7916532/6a8e0ba8-2e04-4daf-9ab4-f89a3418325b)

**Steps to reproduce**

1.  Deploy the TRE using the "Deploy Azure TRE" GitHub Action for the first time (/to an empty Azure environment).
2.  Observe the error above on first run.
3.  Run a second time and observe the error does not occur.

**Azure TRE release version (e.g. v0.14.0 or main)**

main (v0.15.2)

@nasaldrops @jonnyry I'm trying to work out the difference between the two approaches. I ask GitHub Copilot(!)...

---
Compare the following bash:
```
          for i in {1..3}; do
            az acr login --name "${{ secrets.CI_CACHE_ACR_NAME }}" && ec=0 && break || ec=\$? && sleep 10
          done
          # shellcheck disable=SC2242
          (exit ${ec})
```

with 

```
          for i in {1..3}; do
            az acr login --name "${{ secrets.CI_CACHE_ACR_NAME }}" && ec=0 && break || ec="$?" && sleep 10
          done
          # shellcheck disable=SC2242
          (exit "$ec")
```
---

After some explanation is says:

> In this specific context, both scripts will behave the same way as exit codes are integers and do not contain spaces or special characters. However, the second script follows better bash scripting practices.

So #3861 wins ;-) (and doesn't have linting issues). I'll get it merged unless anybody raises an issue. @nasaldrops are we good to close #3853? Thanks.

--- a/.github/workflows/deploy_tre_reusable.yml
+++ b/.github/workflows/deploy_tre_reusable.yml
@@ -253,10 +253,10 @@ jobs:
         run: |
           # shellcheck disable=SC2034,SC2015,SC2125
           for i in {1..3}; do
-            az acr login --name "${{ secrets.CI_CACHE_ACR_NAME }}" && ec=0 && break || ec=\$? && sleep 10
+            az acr login --name "${{ secrets.CI_CACHE_ACR_NAME }}" && ec=0 && break || ec="$?" && sleep 10
           done
           # shellcheck disable=SC2242
-          (exit \$ec)
+          (exit "$ec")
 
       - name: Push cached devcontainer
         run: docker image push ${{ env.CI_CACHE_ACR_URI }}/tredev:${{ inputs.DEVCONTAINER_TAG }}

